File: src/chipschallenge/blockreactions/SuctionbootsTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Inventory.Boots;

/**
 *
 * @author patrik
 */
public class SuctionbootsTo extends BlockReaction {

    private SuctionbootsTo() {
    }
    private static SuctionbootsTo mInstance = null;

    public static synchronized SuctionbootsTo getInstance() {
        if (mInstance == null) {
            mInstance = new SuctionbootsTo();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) throws BlockContainerFullException {
        if (moving.isChip()) {
            takeBoots(Boots.SUCTIONBOOTS);
            standing.destroy();
            //TODO: Play take-item sound
        }
    }

    public boolean canMove(Block moving, Block standing) {
        return moving.isChip() || moving.isBlock();
    }
}

File: ChipsChallenge.Shared/Blockreactions/SuctionbootsTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Boots = Inventory.Boots;
    using Type = Block.Type;

    public class SuctionbootsTo : NoSlipReaction
    {
        private static SuctionbootsTo instance;

        private SuctionbootsTo()
        {
        }

        public static SuctionbootsTo Instance
        {
            get
            {
                lock (typeof(SuctionbootsTo))
                {
                    return instance ?? (instance = new SuctionbootsTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            if (moving.Chip)
            {
                TakeBoots(Boots.SUCTIONBOOTS);
                standing.Replace(CreateBlock(Type.FLOOR));
            }
        }

        public override bool canMove(Block moving, Block standing)
        {
            return moving.Chip || moving.IsBlock();
        }
    }
}